source/melpomene/src/main.rs

```diff
@@ -153,77 +152,110 @@ fn kernel_entry(opts: MelpomeneOptions) {
-                        "{:02}:{:02}:{02}",
-                        time.hour(),
-                        time.minute(),
-                        time.second()
-                    );
-
-                    let date_str =
-                        format!("{:02}/{:02}/{:02}", time.month(), time.day(), time.year());
-
-                    let date_text = Text::new(&date_str, Point::new(28, 35), datetime_style);
-                    let time_text = Text::new(&time_str, Point::new(88, 35), datetime_style);
-
-                    date_text.draw(&mut fc_0).unwrap();
-                    time_text.draw(&mut fc_0).unwrap();
-                    {
-                        let raw_img = fc_0.frame_display().unwrap();
-                        let image = Image::new(&raw_img, Point::new(80, 45));
-                        image.draw(&mut sdisp).unwrap();
-
-                        window.update(&sdisp);
-                        if window.events().any(|e| e == SimulatorEvent::Quit) {
-                            break 'running;
-                        }
+                let style = ring_drawer::BwStyle {
+                    background: Gray8::BLACK,
+                    font: MonoTextStyle::new(&PROFONT_12_POINT, Gray8::WHITE),
+                };
+
+                // At 12-pt font, there is enough room for 16 lines, with 50 chars/line.
+                //
+                // Leave out 4 for the implicit margin of two characters on each gutter.
+                let mut rline = RingLine::<16, 46>::new();
 
-                        Delay::new(Duration::from_secs(1)).await;
-                        sdisp.clear(Gray8::BLACK).unwrap();
+                loop {
+                    // Wait until there is a frame buffer ready. There wouldn't be if we've spammed frames
+                    // before they've been consumed.
+                    let mut fc_0 = loop {
+                        let fc = disp_hdl
+                            .get_framechunk(0, char_y as i32, 400, 240 - char_y)
+                            .await;
+                        if let Some(fc) = fc {
+                            break fc;
+                        } else {
+                            Delay::new(Duration::from_millis(10)).await;
+                        }
+                    };
+                    ring_drawer::drawer_bw(&mut fc_0, &rline, style.clone()).unwrap();
+                    disp_hdl.draw_framechunk(fc_0).await.unwrap();
+
+                    let rgr = p2.consumer().read_grant().await;
+                    for b in rgr.iter() {
+                        match rline.append_local_char(*b) {
+                            Ok(_) => {}
+                            // backspace
+                            Err(_) if *b == 0x7F => {
+                                rline.pop_local_char();
+                            }
+                            Err(_) if *b == b'\n' => {
+                                rline.submit_local_editing();
```

Comment:
it feels a little bit weird to me that we always try to append a character, and then if we get an error, we handle the backspace and newline cases. i feel like it might be a bit easier to follow if we matched on `b`, called `pop_local_char` if it's backspace, called `submit_local_char` if it's `\n`, and called `append_local_char` otherwise? 

this isn't a huge deal to me though, just a thought.

Reply:
Yeah, I guess I know that it is already checking for "good ascii character" inside of "append", so it seemed a little silly to check before then check again inside the library.

We might want more robust input handling in the future (like handling more control characters properly), I might punt on that for now.